### ISSUE TITLE
HLS memcpy: address bits lost when using hls memcpy to copy only one 512bits word

### ISSUE THREAD

**Issue by luyong6** [open-power/snap#263]
Today I found a bug maybe in Vivado HLS
If using 
```
memcpy(buf, (snap_membus *t) (din_gmem + (input_address >> ADDR_RIGHT_SHIFT)), BPERDW);

//ADDR_RIGHT_SHIFT = 6
//BPERDW = 64
```
to transfer only 1 BPERDW, HLS synthesis result shows:
```
signal p_s_reg_151 : STD_LOGIC_VECTOR (63 downto 0);
...
tmp_2_fu_240_p4 <= p_s_reg_151(37 downto 6);
```

p_s_reg_151 (63 downto 0) has the correct full width address, but I don't know why tmp_2_fu_240_p4 will cut higher bits and send the address to m_axi_host_mem_ARADDR, and finally go to ah_cea. 

Simulation on bb x86 server happen to pass, because higher bits (higher than bit37) are zero accidently. 
When running on Power server,  higher bits lost will cause PSL bad context error and timeout.  
(Last time I thought it was because SNAP hadn't developed the part of ah_cch stuff and didn't know it is because of bad address. )

The confusing part is: 
If use this memcpy statement to copy more than 1 word, since it goes to a burst and may be handled by some different logic, the address is tranferred correctly. 
Most of our SNAP example cases are copying data block, so we didn't see the problem in hls_memcopy, hls_hashjoin, hls_intersect, and even the first part of code of hls_bfs. 

I remember in the very beginning, there are functions "read/write_single_word...."  Are they from Xilinx guys?  They are necessary. We need them come back.  And they can perform properly on the address bits.

**Comment by bmesnet** [comment]
Lisa opened problem CR-978469
Waiting for fix

**Comment by bmesnet** [comment]
Received from Lisa@Xilinx that this issue will be fixed in Version 2017.4

**Comment by bmesnet** [comment]
This bug has been confirmed as corrected by Xilinx in release Vivado HLS 2017.4
Closing the issue
